File: examples/showcase/jira_fronting/custom-code-bundles/jira_fronting_go/extensions/jira_live_smoke_test.go
package extensions

import (
	"encoding/base64"
	"encoding/json"
	"net/http"
	"net/url"
	"os"
	"testing"

	"{{ANIP_GO_MODULE_PATH}}/generated"
)

func TestJiraLiveBackendAdapter(t *testing.T) {
	config := readJiraConfig()
	if config == nil {
		t.Skip("Jira credentials are not configured")
	}
	projects := jiraLiveGet(t, config, "/rest/api/3/project/search", map[string]string{"maxResults": "1"})["values"].([]any)
	if len(projects) == 0 {
		t.Fatal("expected at least one Jira project")
	}
	projectKey := projects[0].(map[string]any)["key"].(string)
	issues := jiraLiveGet(t, config, "/rest/api/3/search/jql", map[string]string{"jql": "project = " + projectKey + " ORDER BY updated DESC", "maxResults": "2", "fields": "summary,status,issuetype,project"})["issues"].([]any)
	if len(issues) == 0 {
		t.Fatal("expected at least one Jira issue")
	}
	issueKey := issues[0].(map[string]any)["key"].(string)
	secondIssueKey := issueKey
	if len(issues) > 1 {
		secondIssueKey = issues[1].(map[string]any)["key"].(string)
	}
	search := executeForTest(t, "jira.backlog.search_context", map[string]any{"project_key": projectKey, "query": "test", "limit": 5})
	if search["execution_status"] != "completed" {
		t.Fatalf("expected search completed, got %#v", search)
	}
	issue := executeForTest(t, "jira.issue.get_context", map[string]any{"issue_key": issueKey, "include_comments": true})
	if issue["execution_status"] != "completed" {
		t.Fatalf("expected issue completed, got %#v", issue)
	}
	cases := map[string]map[string]any{
		"jira.incident_bug.prepare":                  {"project_key": projectKey, "summary": "ANIP smoke bug", "description": "Preview only", "severity": "sev3", "labels": []any{"anip-smoke"}},
		"jira.story.prepare":                         {"project_key": projectKey, "summary": "ANIP smoke story", "acceptance_criteria": []any{"Given ANIP", "Then no mutation"}, "priority": "medium"},
		"jira.subtask.prepare":                       {"parent_issue_key": issueKey, "summary": "ANIP smoke subtask", "description": "Preview only"},
		"jira.customer_escalation.comment.prepare":   {"issue_key": issueKey, "comment_purpose": "triage_update", "context": "Preview only", "visibility": "internal"},
		"jira.workflow_transition.request":           {"issue_key": issueKey, "target_status": "To Do", "reason": "Preview only", "comment": "Preview only"},
		"jira.sprint_move.request":                   {"issue_keys": []any{issueKey}, "target_sprint": "preview-sprint", "reason": "Preview only"},
		"jira.assignee_change.request":               {"issue_key": issueKey, "assignee_ref": "preview-account-id", "reason": "Preview only"},
		"jira.issue_link.request":                    {"source_issue_key": issueKey, "target_issue_key": secondIssueKey, "link_type": "Relates", "reason": "Preview only"},
	}
	for id, params := range cases {
		result := executeForTest(t, id, params)
		if result["execution_status"] != "prepared" || result["mutation_performed"] != false {
			t.Fatalf("%s expected prepared without mutation, got %#v", id, result)
		}
	}
}

func executeForTest(t *testing.T, capabilityID string, params map[string]any) map[string]any {
	t.Helper()
	capability := capabilityForTest(t, capabilityID)
	plan := generated.BackendInvocationPlan{SemanticInput: params, AdapterInput: params, BackendInputContract: generated.EffectiveBackendInputContract{Mode: "explicit"}}
	result, err := BackendAdapterInstance.Execute(capability, plan, params, BackendInvocationContext{})
	if err != nil {
		t.Fatal(err)
	}
	return result
}

func capabilityForTest(t *testing.T, capabilityID string) generated.GeneratedCapabilityRuntimeMetadata {
	t.Helper()
	for _, capability := range generated.GeneratedCapabilityMetadata {
		if capability.CapabilityID == capabilityID {
			return capability
		}
	}
	t.Fatalf("missing capability %s", capabilityID)
	return generated.GeneratedCapabilityRuntimeMetadata{}
}

func jiraLiveGet(t *testing.T, config *jiraConfig, path string, query map[string]string) map[string]any {
	t.Helper()
	target, _ := url.Parse(config.BaseURL + path)
	values := target.Query()
	for key, value := range query {
		values.Set(key, value)
	}
	target.RawQuery = values.Encode()
	request, _ := http.NewRequest("GET", target.String(), nil)
	request.Header.Set("Accept", "application/json")
	request.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(os.Getenv("JIRA_EMAIL")+":"+os.Getenv("JIRA_API_TOKEN"))))
	response, err := http.DefaultClient.Do(request)
	if err != nil {
		t.Fatal(err)
	}
	defer response.Body.Close()
	var payload map[string]any
	if err := json.NewDecoder(response.Body).Decode(&payload); err != nil {
		t.Fatal(err)
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		t.Fatalf("Jira GET %s failed: %#v", path, payload)
	}
	return payload
}

File: examples/showcase/jira_fronting/custom-code-bundles/jira_fronting_typescript/tests/jira-live-smoke.test.ts
import { describe, expect, it } from "vitest";
import { backendAdapter } from "../src/runtime/backend-adapter.js";
import { generatedCapabilityMetadata, type BackendInvocationPlan, type GeneratedCapabilityRuntimeMetadata } from "../src/generated/runtime-target.js";

const configured = Boolean(process.env.JIRA_BASE_URL && process.env.JIRA_EMAIL && process.env.JIRA_API_TOKEN);

function capability(id: string): GeneratedCapabilityRuntimeMetadata {
  const found = generatedCapabilityMetadata.find((item) => item.capability_id === id);
  if (!found) throw new Error(`Missing capability ${id}`);
  return found;
}

function plan(parameters: Record<string, unknown>): BackendInvocationPlan {
  return {
    selected_binding: null,
    semantic_input: parameters,
    adapter_input: parameters,
    backend_input_contract: { mode: "explicit", required: [], optional: [] },
    unresolved_required_backend_inputs: [],
  };
}

async function jiraGet(path: string, query: Record<string, string> = {}): Promise<Record<string, unknown>> {
  const url = new URL(`${process.env.JIRA_BASE_URL!.replace(/\/+$/, "")}${path}`);
  for (const [key, value] of Object.entries(query)) url.searchParams.set(key, value);
  const response = await fetch(url, {
    headers: {
      Accept: "application/json",
      Authorization: `Basic ${Buffer.from(`${process.env.JIRA_EMAIL}:${process.env.JIRA_API_TOKEN}`).toString("base64")}`,
    },
  });
  expect(response.ok).toBe(true);
  return await response.json() as Record<string, unknown>;
}

describe.skipIf(!configured)("Jira live backend adapter", () => {
  it("executes bounded reads and prepares governed previews without mutation", async () => {
    const projects = ((await jiraGet("/rest/api/3/project/search", { maxResults: "1" })).values ?? []) as Record<string, unknown>[];
    expect(projects.length).toBeGreaterThan(0);
    const projectKey = String(projects[0].key);
    const issues = ((await jiraGet("/rest/api/3/search/jql", { jql: `project = ${projectKey} ORDER BY updated DESC`, maxResults: "2", fields: "summary,status,issuetype,project" })).issues ?? []) as Record<string, unknown>[];
    expect(issues.length).toBeGreaterThan(0);
    const issueKey = String(issues[0].key);
    const secondIssueKey = String((issues[1] ?? issues[0]).key);

    const search = await backendAdapter.execute(capability("jira.backlog.search_context"), plan({ project_key: projectKey, query: "test", limit: 5 }), { project_key: projectKey, query: "test", limit: 5 }, {});
    expect(search.execution_status).toBe("completed");

    const issue = await backendAdapter.execute(capability("jira.issue.get_context"), plan({ issue_key: issueKey, include_comments: true }), { issue_key: issueKey, include_comments: true }, {});
    expect(issue.execution_status).toBe("completed");

    const previews: Array<[string, Record<string, unknown>]> = [
      ["jira.incident_bug.prepare", { project_key: projectKey, summary: "ANIP smoke bug", description: "Preview only", severity: "sev3", labels: ["anip-smoke"] }],
      ["jira.story.prepare", { project_key: projectKey, summary: "ANIP smoke story", acceptance_criteria: ["Given ANIP", "Then no mutation"], priority: "medium" }],
      ["jira.subtask.prepare", { parent_issue_key: issueKey, summary: "ANIP smoke subtask", description: "Preview only" }],
      ["jira.customer_escalation.comment.prepare", { issue_key: issueKey, comment_purpose: "triage_update", context: "Preview only", visibility: "internal" }],
      ["jira.workflow_transition.request", { issue_key: issueKey, target_status: "To Do", reason: "Preview only", comment: "Preview only" }],
      ["jira.sprint_move.request", { issue_keys: [issueKey], target_sprint: "preview-sprint", reason: "Preview only" }],
      ["jira.assignee_change.request", { issue_key: issueKey, assignee_ref: "preview-account-id", reason: "Preview only" }],
      ["jira.issue_link.request", { source_issue_key: issueKey, target_issue_key: secondIssueKey, link_type: "Relates", reason: "Preview only" }],
    ];
    for (const [id, parameters] of previews) {
      const result = await backendAdapter.execute(capability(id), plan(parameters), parameters, {});
      expect(result.execution_status).toBe("prepared");
      expect(result.mutation_performed).toBe(false);
    }
  });
});
